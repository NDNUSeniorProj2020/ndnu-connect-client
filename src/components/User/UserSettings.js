import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { message } from 'antd';

import './User.css'
import { hasToken } from '../../actions/auth/authenticationActions';
import { updateUser } from '../../actions/users/userActions';
import UserSettingsForm from './UserSettingsForm';

export function ConnectedUserSettings({ user, errors, updated, success, hasToken, updateUser }) {
  useEffect(() => {
    hasToken();
  }, [hasToken]);

  const update = (user = {}) => updateUser(user);

  return (
    <div>
      { updated ? message.success('Successfully updated your account!', 10) : null }
      { Object.keys(errors).length > 0 && errors.msg.length > 0 ? message.error(errors.msg, 10) : null }
      <h1>Settings</h1>
      <UserSettingsForm user={user} updateUser={update} />
    </div>
  );
}

ConnectedUserSettings.propTypes = {
  user: PropTypes.object,
  errors: PropTypes.object,
  success: PropTypes.bool,
  hasToken: PropTypes.func,
  updateUser: PropTypes.func
};
ConnectedUserSettings.defaultProps = {
  user: {},
  errors: {},
  success: false,
  updated: false,
  hasToken: f => f,
  updateUser: f => f
};

const mapStateToProps = ({ authReducer, userReducer }) => ({
  user: authReducer.user,
  updated: userReducer.updated,
  errors: authReducer.errors,
  success: authReducer.success
});
const UserSettings = connect(mapStateToProps, { hasToken, updateUser })(ConnectedUserSettings);

export default UserSettings;
