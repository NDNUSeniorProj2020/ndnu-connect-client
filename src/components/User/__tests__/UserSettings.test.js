import React from 'react';
import { shallow, mount } from 'enzyme';

import { ConnectedUserSettings } from '../UserSettings';

const user = {
  "id": 1,
  "email": "JeffWorker@gmail.com",
  "first_name": "John",
  "last_name": "Worker",
  "phone_number": "",
  "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwiZXhwdCI6NX0._hJH_2eZrFsxawxJKdm2AlZkWgvD6ADWO3xQW2kiui8",
  "graduated": true,
  "year_graduated": 2017,
  "major": "Computer Science",
  "company": "Google Inc.",
  "job_title": "Software Engineer",
  "about": null
};
const errors = {
  msg: 'Failed.'
};

describe('tests for UserSettings components', () => {
  describe('ConnectedUserSettings tests', () => {
    describe('snapshot tests', () => {
      it('renders without crashing', () => {
        const tree = shallow(<ConnectedUserSettings />);
        expect(tree).toMatchSnapshot();
      });
    });

    describe('unit tests', () => {
      it('has default props', () => {
        const { user, errors, success, hasToken } = ConnectedUserSettings.defaultProps;
        expect(user).toEqual({});
        expect(errors).toEqual({});
        expect(success).toEqual(false);
        expect(hasToken('f')).toEqual('f');
      });
    });

    describe('integration tests', () => {
      // Test setup
      let props;
      let wrapper;
      let useEffect;

      beforeEach(() => {
        useEffect = jest.spyOn(React, 'useEffect').mockImplementation(f => f());
        props = { hasToken: jest.fn().mockResolvedValue(user) }
        wrapper = mount(<ConnectedUserSettings {...props} />)
      });

      afterEach(() => jest.clearAllMocks());

      it('fetches user on load', () => {
        expect(props.hasToken).toHaveBeenCalled();
      });

      it('throws an error if something fails', () => {
        // If errors are present, an error popup message will be rendered
        mount(<ConnectedUserSettings {...props} errors={errors} />);
      });
    });
  });
});
