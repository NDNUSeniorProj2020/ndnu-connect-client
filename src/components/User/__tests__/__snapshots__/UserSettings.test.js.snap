// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for UserSettings components ConnectedUserSettings tests snapshot tests renders without crashing 1`] = `ShallowWrapper {}`;
